### PR TITLE
Check for Hyper-V 2016 with UEFI, configure VM network adapter, and log related soft failure.

### PR DESCRIPTION
**Pull request by RoyCai7**
Check for Hyper-V 2016 with UEFI, configure VM network adapter, and log related soft failure.

1. Added hyperv_utils.pm for Hyper-V command execution.
2. Disable network card when create VM, record soft failure. (module: [bootloader_hyperv](https://openqa.oqa.prg2.suse.org/tests/15702722#step/bootloader_hyperv/29))
3. Enable network card when scc registration, record soft failure (module: 
[scc_registration](https://openqa.oqa.prg2.suse.org/tests/15702722#step/scc_registration/4))
4. Disable network card when the first boot, record soft failure (module: [reboot_after_installation](https://openqa.oqa.prg2.suse.org/tests/15702722#step/reboot_after_installation/5))
5. Enable network card when function test, record soft failure (module: [system_prepare](https://openqa.oqa.prg2.suse.org/tests/15702722#step/system_prepare/16))

- Related ticket: https://progress.opensuse.org/issues/151645
- Needles: 
     - scc_registration-net-dhcp-20241015 
     - scc_registration-net-setup-20241015 
     - scc_registration-net-20241015 
     - scc_registration-net-dhcp-20241016 
     - scc_registration-net-setup-20241016 
     - scc-registration-net-selected-20241016
     - scc_registration-net-20241016 

- Verification run:  
     - [online_upgrade_sles15sp6_hyperv@svirt-hyperv2016](https://openqa.oqa.prg2.suse.org/tests/15702724)
     - [default_install_svirt@svirt-hyperv2016-uefi](https://openqa.oqa.prg2.suse.org/tests/15702723)
     - [textmode_svirt@svirt-hyperv2016-uefi](https://openqa.oqa.prg2.suse.org/tests/15702722)
     - [textmode_svirt@svirt-hyperv2022-uefi](https://openqa.oqa.prg2.suse.org/tests/15702746)
     - [default_install_svirt@svirt-hyperv2022-uefi](https://openqa.oqa.prg2.suse.org/tests/15702747)

### CODE REVIEW

File: lib/virt_autotest/hyperv_utils.pm [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# SUSE's openQA tests
+#
+# Copyright 2024 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Summary: Utilities for executing commands on Hyper-V hosts, allowing interaction
+#          with Hyper-V plugins and managing the Hyper-V environment within openQA tests.
+# Maintainer: qe-virt@suse.de
```
Comment (waynechen55) [inline lib/virt_autotest/hyperv_utils.pm:8]: Please also add your name after qe-virt.

File: lib/virt_autotest/hyperv_utils.pm [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+
+use base 'Exporter';
+use Exporter;
+use base 'installbasetest';
+use testapi;
+use utils;
+use strict;
+use warnings;
+use File::Basename;
```
Comment (waynechen55) [inline lib/virt_autotest/hyperv_utils.pm:20]: Remove unused imported modules.

File: tests/console/system_prepare.pm [hunk 3]
```diff
@@ -35,6 +36,17 @@ sub run {
 
     prepare_serial_console;
 
+    # This code checks if the environment is Hyper-V 2016 with UEFI. If true,
+    # it adds a new VM network adapter connected to a virtual switch and logs a known UEFI boot issue.
+    if (check_var('HYPERV_VERSION', '2016') && get_var('UEFI')) {
```
Comment (waynechen55) [inline tests/console/system_prepare.pm:41]: Use function ```is_uefi_boot```

File: tests/console/system_prepare.pm [hunk 4]
```diff
@@ -35,6 +36,17 @@ sub run {
...
+        my $virsh_instance = get_var("VIRSH_INSTANCE");
```
Comment (waynechen55) [inline tests/console/system_prepare.pm:42]: get_required_var ???

File: tests/installation/bootloader_hyperv.pm [hunk 5]
```diff
@@ -15,15 +15,7 @@ use strict;
...
-    my ($cmd, $args) = @_;
-    $args->{ignore_return_code} ||= 0;
-    my $ret = console('svirt')->run_cmd($cmd);
-    diag "Command on Hyper-V returned: $ret";
-    die 'Command on Hyper-V failed' unless ($args->{ignore_return_code} || !$ret);
-    return $ret;
-}
+use virt_autotest::hyperv_utils 'hyperv_cmd';
 
 sub hyperv_cmd_with_retry {
```
Comment (waynechen55) [inline tests/installation/bootloader_hyperv.pm:20]: hyperv_cmd_with_retry belongs to hyperv_utils.pm ???

File: tests/installation/reboot_after_installation.pm [hunk 6]
```diff
@@ -45,6 +46,18 @@ sub run {
...
+    # If the current test is running on a Hyper-V host with version 2016 and UEFI is enabled:
+    # - Get the VM instance name stored in the VIRSH_INSTANCE variable
+    # - Run a PowerShell command to remove the network adapter from the VM (openQA-SUT) in Hyper-V
+    # - Record a soft failure with a reference to bug bsc#1217800, which is related to UEFI boot issues
+    #   encountered on Hyper-V guests using different ISO builds on bare-metal Windows Server 2016.
+    if (check_var('HYPERV_VERSION', '2016') && get_var('UEFI')) {
+        my $virsh_instance = get_var("VIRSH_INSTANCE");
+        hyperv_cmd("powershell -Command Remove-VMNetworkAdapter -VMName openQA-SUT-${virsh_instance}");
+        record_soft_failure('bsc#1217800 - [Baremetal windows server 2016][guest VM UEFI]UEFI Boot Issues with Different Build ISOs on Hyper-V Guests');
+    }
```
Comment (waynechen55) [inline tests/installation/reboot_after_installation.pm:59]: Already saw multiple occurrence of this block. Any way to wrap and reused ???

File: lib/virt_autotest/hyperv_utils.pm (hunk 1 above)
Comment (RoyCai7) [reply lib/virt_autotest/hyperv_utils.pm:8]: Added

File: lib/virt_autotest/hyperv_utils.pm (hunk 2 above)
Comment (RoyCai7) [reply lib/virt_autotest/hyperv_utils.pm:20]: removed basename

File: tests/console/system_prepare.pm (hunk 3 above)
Comment (RoyCai7) [reply tests/console/system_prepare.pm:41]: Done

File: tests/console/system_prepare.pm (hunk 4 above)
Comment (RoyCai7) [reply tests/console/system_prepare.pm:42]: replaced

File: tests/installation/bootloader_hyperv.pm (hunk 5 above)
Comment (RoyCai7) [reply tests/installation/bootloader_hyperv.pm:20]: I don't use it this time, if next time use it in other place, will add it to hyperv_utils.pm.

File: tests/installation/reboot_after_installation.pm (hunk 6 above)
Comment (RoyCai7) [reply tests/installation/reboot_after_installation.pm:59]: These functions are not frequently used, and they are different from each other. I don't think it's necessary to integrate them into one function, as it feels more complex.

**Comment by martinsmarcelo** [comment]
LGTM for me

File: tests/installation/bootloader_hyperv.pm (hunk 5 above)
Comment (waynechen55) [reply tests/installation/bootloader_hyperv.pm:20]: Why next time ??? It looks to me it should be moved to hyperv_utils.pm obviously.

Comment (RoyCai7) [reply tests/installation/bootloader_hyperv.pm]: all right, I will add it to utils.

Comment (RoyCai7) [reply tests/installation/bootloader_hyperv.pm]: Done

**Comment by RoyCai7** [comment]
Thank you, Wayne, for all the great advice.

**Comment by Julie-CAO** [comment]
Shall we remove win2016 uefi hyperV tests per https://bugzilla.suse.com/show_bug.cgi?id=1217800#c34?

**Comment by alice-suse** [comment]
> @alice-suse Shall we remove win2016 uefi hyperV tests per https://bugzilla.suse.com/show_bug.cgi?id=1217800#c34?

Hi @Julie-CAO  , I already talked with Roy earlier. We need to keep it because the product will still live for several years. So request Roy to add the workaround .

**Pull request merged**